velox/serializers/UnsafeRowSerializer.cpp

```diff
@@ -0,0 +1,158 @@
+ * limitations under the License.
+ */
+#include "velox/serializers/UnsafeRowSerializer.h"
+#include "velox/row/UnsafeRowDeserializer.h"
+#include "velox/row/UnsafeRowDynamicSerializer.h"
+
+namespace facebook::velox::serializer::spark {
+
+void UnsafeRowVectorSerde::estimateSerializedSize(
+    VectorPtr vector,
+    const folly::Range<const IndexRange*>& ranges,
+    vector_size_t** sizes) {
+  for (auto& range : ranges) {
+    for (auto i = range.begin; i < range.begin + range.size; ++i) {
+      size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSize(
+          vector->type(), vector, i);
+      (*sizes)[i] += rowSize;
+    }
+  }
+}
+
+namespace {
+class UnsafeRowVectorSerializer : public VectorSerializer {
+ public:
+  UnsafeRowVectorSerializer(StreamArena* streamArena)
+      : streamArena_{streamArena} {}
+
+  void append(
+      RowVectorPtr vector,
+      const folly::Range<const IndexRange*>& ranges) override {
+    size_t totalSize = 0;
+    for (auto& range : ranges) {
+      for (auto i = range.begin; i < range.begin + range.size; ++i) {
+        size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSize(
+            vector->type(), vector, i);
+        totalSize += rowSize + sizeof(size_t);
+      }
+    }
+
+    if (totalSize == 0) {
+      return;
+    }
+
+    // Try to allocate 'totalSize' bytes, but handle cases where allocated size
+    // is less.
+    ByteRange byteRange;
+    streamArena_->newRange(totalSize, &byteRange);
+    VELOX_CHECK_EQ(byteRange.position, 0);
+
+    size_t offset = 0;
+    for (auto& range : ranges) {
+      for (auto i = range.begin; i < range.begin + range.size; ++i) {
+        // Write row data.
+        size_t rowSize = velox::row::UnsafeRowDynamicSerializer::getSize(
+            vector->type(), vector, i);
+
+        // Check if there is enough space in the buffer. If not, allocate a new
+        // buffer.
+        if (byteRange.size < offset + rowSize + sizeof(size_t)) {
+          byteRange.position = offset;
+          buffers_.push_back(byteRange);
+
+          totalSize -= offset;
+          streamArena_->newRange(totalSize, &byteRange);
+          offset = 0;
+
+          // Give up if cannot allocate enough contiguous memory for a single
+          // row.
+          VELOX_CHECK_LE(
+              rowSize + sizeof(size_t),
+              byteRange.size,
+              "Single row is too large for serialization");
+        }
+
+        auto size = velox::row::UnsafeRowDynamicSerializer::serialize(
+                        vector->type(),
+                        vector,
+                        (char*)byteRange.buffer + offset + sizeof(size_t),
+                        i)
+                        .value_or(0);
+
+        // Sanity check.
+        VELOX_CHECK_EQ(rowSize, size);
+
+        // Write raw size.
+        *(size_t*)(byteRange.buffer + offset) = size;
+
+        offset += sizeof(size_t) + size;
+      }
+    }
+    byteRange.position = offset;
```

Comment:
In the normal situation (with no overflow) at this point offset will be equal to ```totalSize``` and ```buffers_``` will have only one entry. Just making sure that's the case.

Reply:
It is not always the case that we are able to allocate totlaSize in one chunk. Hence, in some cases there will be multiple buffers.

Reply:
Got it.